### PR TITLE
Replace Socket, ServerSocket, and HttpServer usages in tests with mocksocket versions

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #22116. In order to remove SocketPermissions from core we need to isolate some code that can be given SocketPermissions. The mocksocket jar includes wrapped versions of Socket and SocketServer. This is one step towards migrating permissions out of core.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This is currently blocked from being merged on a release of the [mocksocket](https://github.com/elastic/mocksocket) project.

I should get to today or tomorrow.

**Review by s1monw — APPROVED** [review]
this PR LGTM provided MockSocket is available publicially

File: modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoderTests.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.transport.netty4;
 
+import org.elasticsearch.MockSocket;
```
Comment (jasontedor) [inline modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoderTests.java]: My only comment is whether or not this should be in another package like `org.elasticsearch.mocksocket`? In general we should avoid split packages, and in a JDK 9 world, split packages are a thing that we are definitely not going to want to do. What do you think @s1monw?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one comment.

File: modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoderTests.java (hunk 1 above)
Comment (jasontedor) [inline modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoderTests.java]: Split packages are packages that live in two jars. Yes, it needs to be fixed in mocksocket but this seemed a good place to raise the conversation.

**Comment by tbrooks8** [comment]
I made a few changes to this PR:

1. I placed the classes in a mocksocket package to attempt to address @jasontedor's concern.
2. I added wrappers for HttpServer and HttpsServer. This addresses the problem where those classes do not wrap `accept()` in `doPrivilege()`.

The mocksocket project is in MavenCentral. So once this passes an updated review, it should be good to merge in.

**Comment by tbrooks8** [comment]
I updated this PR with a new version of mocksocket that adds a [method](https://github.com/elastic/mocksocket/blob/master/src/main/java/org/elasticsearch/mocksocket/MockServerSocket.java#L77) needed for "connect" permissions.

And I rebased it on current master.

**Comment by s1monw** [comment]
LGTM

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**